libs/megaparse/src/api/megaparse_service.py

```diff
@@ -0,0 +1,190 @@
+import io
+import os
+from typing import Optional
+
+import httpx
+import nats
+import psutil
+from langchain_anthropic import ChatAnthropic
+from langchain_community.document_loaders.url_playwright import (
+    UnstructuredHtmlEvaluator,
+)
+from langchain_openai import ChatOpenAI
+from loguru import logger
+from megaparse import MegaParse
+from megaparse.parser.builder import ParserBuilder
+from megaparse.parser.unstructured_parser import UnstructuredParser
+from megaparse_schema.extensions import FileExtension
+from megaparse_schema.mp_exceptions import ModelNotSupported, ParsingException
+from megaparse_schema.mp_inputs import (
+    FileInput,
+    MPInput,
+    ParseFileConfig,
+    ParseFileInput,
+    ParseUrlInput,
+)
+from megaparse_schema.mp_outputs import MPErrorType, MPOutput, MPOutputType, ParseError
+from megaparse_schema.parser_config import StrategyEnum
+from nats.aio.client import Client
+from nats.aio.msg import Msg
+from playwright.async_api import Browser, async_playwright
+
+from api.utils.load_ssl import load_ssl_cxt
+
+
+def _check_free_memory() -> bool:
+    """Reject traffic when free memory is below minimum (default 2GB)."""
+    mem = psutil.virtual_memory()
+    memory_free_minimum = int(os.environ.get("MEMORY_FREE_MINIMUM_MB", 2048))
+
+    if mem.available <= memory_free_minimum * 1024 * 1024:
+        return False
+    return True
+
+
+class MegaParseService:
+    def __init__(
+        self, browser: Browser, html_remove_selectors: Optional[list[str]] = None
+    ) -> None:
+        self.parser_builder = ParserBuilder()
+        self.browser = browser
+        self.httpx_client = None
+        self.html_evaluator = UnstructuredHtmlEvaluator(html_remove_selectors)
+
+    def _init_model(self, model_name: str) -> ChatOpenAI | ChatAnthropic:
+        if model_name.startswith("gpt"):
+            model = ChatOpenAI(model=model_name, api_key=os.getenv("OPENAI_API_KEY"))  # type: ignore
+        elif model_name.startswith("claude"):
+            model = ChatAnthropic(
+                model_name=model_name,
+                api_key=os.getenv("ANTHROPIC_API_KEY"),  # type: ignore
+                timeout=60,
+                stop=None,
+            )
+        else:
+            raise ModelNotSupported
+        return model
+
+    async def parse_file(
+        self, file: FileInput, parser_config: ParseFileConfig
+    ) -> MPOutput:
+        logger.info(f"Parsing file {file.file_name} using config {parser_config}")
+        if not _check_free_memory():
+            return MPOutput(
+                output_type=MPOutputType.PARSE_ERR,
+                err=ParseError(
+                    mp_err_code=MPErrorType.MEMORY_LIMIT,
+                    message="high memory pressure on server",
+                ),
+                result=None,
+            )
+        try:
+            if parser_config.llm_model_name and parser_config.check_table:
+                model = self._init_model(parser_config.llm_model_name)
```

Comment:
I feel like model is never given to the parser since it is not added in the parser_config so in self.parser_builder.build() self.model will always be None even if you give a llm_model_name